### ISSUE TITLE
Issues w/ image-based subtitles and/or audio downmixing

### ISSUE THREAD

**Issue by kmescher** [jellyfin/jellyfin-androidtv#4193]
### Describe the bug

If these are duplicate issues, please point me in that direction; I haven't found anything (current) that matches what I'm seeing.

To start off with, all my video files are rips directly off my physical media.  I do very little (or no) conversion.  So with audio tracks, they are the same as on the media (format & number of channels).  Same with subtitle tracks.

This means that for an average DVD Rip, I will have either an AC3 or a DTS audio track, and a VOBSUB subtitle track.
For a Blu-Ray rip, audio tracks are far more varied (add TrueHD, DTSHD-MA, etc), and PGS subtitle tracks.

I store everything in MKV (most versatile).

I see no need to transcode at this point; I believe the player should be able to handle this (somewhat basic) set of tracks.  And as my server is running on a Pi, transcoding is a bit troublesome anyway.

When I attempt to play a DVD rip, I receive an reason of "SubtitleCodec not supported in the log" when trying to direct play.  The actual playback just sits at a black screen (no error popups).

The profile says that dvdsub is supported.  Is there something I need to do to indicate that the VOBSUB's are actually dvdsubs, or is the player lying about its capabilities?  Does Exoplayer even support vobsub's?

When I attempt to play a BluRay rip, I don't receive any errors or reasons why it can't DirectPlay, but the playback (just as above) just sits at a black screen.

I've seen various posts over the past few years stating that image-based subtitle formats aren't supported, then are supported, but there's no clear documentation (like on the Jellyfin web site) if they truly a) are supported by the AndroidTV ExoPlayer, and b) are reasonably expected to work correctly.  Some clarity would be helpful here for people who just want to archive their own media libraries.

Now to an audio question:

I select "downmix to stereo" as I don't have a surround setup at the moment.  I would have expected that option to instruct the player to do the downmixing.  Evidently, that option seems to force the player to only allow 2 channels streams to directplay; anything with more channels (like all of my library) refuse to directplay.  Not surprisingly, I get an error of AudioCodecNotSupported, and because the number of channels is greater than 2.

My question is this: Does ExoPlayer support a stereo downmix locally, or does it rely on the server to accomplish this?  If it relies on the server, it would probably be a good idea to change the option text to reflect this (Like "Request server downmix to stereo").  The current text implies (to me) that the player will be doing it on its own (which would then mean that it doesn't care what codec it's sent, and should be able to directPlay).

I read that the libVLC option was removed a few versions ago (the reasons given weren't very descriptive).  I was a bit dismayed, as this worked beautifully (aside from some sync timings which I was able to correct easily enough).  I would have assumed it would have stayed an option until ExoPlayer could at least match its capabilities.  This does not seem to be the case.  I now have to use VLC as an external player, as it works very well (aside from saved subtitle language, no resume support, and it constantly asking if I should mark the video as watched).  While VLC is a terrific player, using it seems like an unfortunate hack, and wouldn't you want to support & present the most capable player as the default?  ExoPlayer seems unusually limited for even the most straightforward case (playing physical media rips), and there's not a lot of official documentation on how to actually support this.  It's a good thing I explore the options.

### Logs

For a DVD Rip using ExoPlayer w/ downmix option turned on:

[DBG] DirectPlay Result for Profile: "AndroidTV-ExoPlayer", Path: "/media/movies/movies/The_Abyss/The_Abyss.mkv", PlayMethod: null, AudioStreamIndex: 1, SubtitleStreamIndex: 2, Reasons: AudioCodecNotSupported, SubtitleCodecNotSupported

For a DVD Rip using ExoPlayer w/ downmix option turned off:

[DBG] DirectPlay Result for Profile: "AndroidTV-ExoPlayer", Path: "/media/movies/movies/The_Abyss/The_Abyss.mkv", PlayMethod: null, AudioStreamIndex: 1, SubtitleStreamIndex: 2, Reasons: SubtitleCodecNotSupported

For a BluRay rip using ExoPlayer w/ downmix option turned off:

[DBG] DirectPlay Result for Profile: "AndroidTV-ExoPlayer", Path: "/media/movies/movies/Apollo 13/Apollo 13.mkv", PlayMethod: DirectPlay, AudioStreamIndex: 1, SubtitleStreamIndex: 4, Reasons: 0

### FFmpeg logs

```shell

```

### Media info of the file

```shell

```

### Application version

0.17.9

### Device information

Tivo Steam 4K

### Android version

Android 10 (I think)

### Jellyfin server version

10.10

**Comment by nielsvanvelzen** [comment]
Did you try with the 0.18 beta?

**Comment by rickysixx** [comment]
ExoPlayer should be able to play surround audio tracks also on non-surround audio hardware without the need of audio transcoding on the server. You should not need to turn on "Downmix to stereo". If your TV does not support DTS/Dolby True HD, you can try to update the app to 0.18 beta and enable "Prefer FFmpeg for audio decoding" option under Developer settings.

If you are direct playing the media and your TV supports PGS subtitles, they should work fine with ExoPlayer. If instead you have "Downmix to stereo" enabled, and therefore transcoding audio, they will not be shown unless you also transcode the video to burn them in. I don't know about VOBSUB subtitles, but I think they need to be burned-in too (if unsupported by the device) since they also are picture-based.

If your TV does not support PGS/VOBSUB subtitles and you want to avoid video transcoding to burn them in (which can be very hardware-demanding for blu-ray/4K rips and will cause quality loss), you may consider using [SubtitleEdit](https://github.com/SubtitleEdit/subtitleedit) (or similar tools) to convert your PGS/VOBSUB subtitle track to SRT format.

**Comment by kmescher** [comment]
Sorry for the delay.  I just downloaded 0.18.3 beta, and gave it a try.

Just for review, I have a Tivo Stream 4K.

Here is my experience for DVDs:
- When I first started, I got a black screen (no video).  I had to go into
the developer options and select "Enable FFMpeg for audio" option.  One
track was DTS, the other was Dolby Digital.
- When I attempted to turn on the subtitles (now displayed as DVDSUB
(instead of VOBSUB)), the selection did not take effect.  I saw no
subtitles, and when I went to turn them on (again), the checkbox was still
on "None". I believe that they used to show up as VOBSUB.
- Audio was set to "Direct", and played back without issue. (after setting
the FFMpeg option as stated above).

I tried two different DVD rips; same behavior between the two.

Blu-Ray performance was better:
- I was able to start a file without issue (I did this check after the DVD
video check, so the FFMpeg setting may have been required here as well)
- The PGS subtitles worked correctly.

I tried with two different Blu-Ray rips, and they both performed correctly.

In conclusion, the "enable FFMpeg" option seemed to be required, and
VOBSUB's do not work (showing up as "DVDSUB")

Thanks for helping out and let me know if there is any additional
troubleshooting that you would like me to do.

**Comment by rickysixx** [comment]
For VOBSUB/DVDSUB check on the server if the option "Allow video playback that requires transcoding" is enabled for your user (server > dashboard > users > your user > media playback). If they are not shown you are probably missing this option.

**Comment by kmescher** [comment]
Riccardo,

I have tested that option previously, but the whole point was to not
transcode.  I do not believe that it should be necessary for a simple
capability.  As VLC can render DVD subtitles perfectly (and has been able
to do so for a very long time), I would have expected that capability to
exist in Exoplayer as well.

I was using libVLC quite successfully (with no transcoding necessary) until
that feature was removed.

**Comment by rickysixx** [comment]
I get your point, but VOBSUB and DVDSUB are not supported by ExoPlayer, which is mantained by Google. This will not change until someone opens a PR in their repo to implement this capability.

If you want to avoid transcoding your only option in the Jellyfin app is to use an external player.
